lib/features/scanner/scanner_bloc.dart

```diff
@@ -0,0 +1,185 @@
+
+import 'package:barcode_scan/barcode_scan.dart';
+import 'package:flutter/cupertino.dart';
+import 'package:permission_handler/permission_handler.dart';
+import 'package:rxdart/rxdart.dart';
+import 'package:seeds/features/scanner/scanner_service.dart';
+import 'package:seeds/providers/services/permission_service.dart';
+import 'package:seeds/utils/bloc/cmd_common.dart';
+
+abstract class ScannerCmd extends Cmd {}
+class QueryCameraPermissionCmd extends ScannerCmd {}
+class RequestCameraPermissionCmd extends ScannerCmd {}
+class OpenSettingsCmd extends ScannerCmd {}
+class StartScannerCmd extends ScannerCmd {}
+class UpdateTextCmd extends ScannerCmd {
+  final TextEditingController textController;
+  final String text;
+  UpdateTextCmd(this.textController, this.text);
+}
+class SetTextControllerCmd extends ScannerCmd {
+  final TextEditingController textController;
+  final ScanContentType updateOnType;
+  SetTextControllerCmd(this.textController, this.updateOnType);
+
+  @override
+  bool operator ==(Object other) => identical(this, other) ||
+    other is SetTextControllerCmd &&
+    runtimeType == other.runtimeType &&
+    textController == other.textController &&
+    updateOnType == other.updateOnType;
+
+  @override int get hashCode => textController.hashCode ^ updateOnType.hashCode;
+}
+
+class ScannerBloc {
+
+  ScannerService _scannerService;
+  PermissionService _permissionService;
+  final _available = BehaviorSubject<ScannerAvailable>();
+  final _scanResult = BehaviorSubject<ScanResult>();
+  final _status = BehaviorSubject<ScanStatus>();
+  final _execute = PublishSubject<ScannerCmd>();
+  final _textController = BehaviorSubject<SetTextControllerCmd>();
+  final _inviteCode = BehaviorSubject<String>();
+
+  Stream<ScannerAvailable> get available => _available.stream;
+  Stream<ScannedData> get data => _scanResult.stream
+    .where((result) => _scannerService.statusFromResult(result) == ScanStatus.successful)
+    .map((result) => ScannedData(result.rawContent, _scannerService.contentTypeOf(result.rawContent)));
+  Stream<ScanStatus> get status => _scanResult.stream.map(_scannerService.statusFromResult);
+  Stream<String> get inviteCode => _inviteCode.stream;
+  Stream<SetTextControllerCmd> get textController => _textController.stream.distinct();
+  Function(ScannerCmd) get execute => _execute.add;
+
+  ScannerBloc() {
+    _execute.listen(_executeCommand);
+    _initScanResultStatus();
+    _initUpdateTextController();
```

Comment:
Ok since this is my chance to ask about blocs

Why are there these init methods - why are there 3 of them instead of 1, why do they look like this?

Reply:
That's just my way of putting a label on each of these code blocks in the init method. 

Consider this constructor doing exactly the same thing:
```
  ScannerBloc() {
    _execute.listen(_executeCommand);

    _scanResult.stream
      .map((scanResult) => _scannerService.statusFromResult(scanResult))
      .listen(_status.add);

    CombineLatestStream
      .combine2(_textController.distinct(), inviteCode, (SetTextControllerCmd ctrl, code) {
      return ctrl.updateOnType == ScanContentType.inviteCode ? UpdateTextCmd(ctrl.textController, code) : null;
    }).where((cmd) => cmd != null)
      .listen(execute);

    data
      .where((data) => data.type == ScanContentType.inviteUrl)
      .flatMap((data) => _scannerService.decodeInviteUrl(data.data))
      .map((uri) => uri.queryParameters["inviteMnemonic"])
      .listen(_inviteCode.add);

    data
      .where((data) => data.type == ScanContentType.inviteCode)
      .map((data) => data.data)
      .listen(_inviteCode.add);
  }
```

Reply:
Btw, this has of course nothing to do with BLOC. It's just me wanting to make the code more readable.

Reply:
That makes a lot of sense thanks